### ISSUE TITLE
調査: ベンチマーク測定ノイズの実態と採用閾値の妥当性検証

### ISSUE THREAD

**Issue by flipslidersand** [flipslidersand-labs/forge#5]
## 背景

ADR-004 では「3% 以上の改善のみ採用、p80 < p20/1.03 で判定」と決めたが、
この閾値が自分の GPU 環境のノイズレベルと合っているかを実測で確認する必要がある。
閾値が緩すぎると誤採用、厳しすぎると真の改善を見逃す。

---

### 1. 同じカーネルを 2 回測定して揺れ幅を確認

```python
import torch, statistics

def measure(fn, *args, warmup=25, repeat=200):
    for _ in range(warmup):
        fn(*args)
    torch.cuda.synchronize()
    samples = []
    for _ in range(repeat):
        start = torch.cuda.Event(enable_timing=True)
        end = torch.cuda.Event(enable_timing=True)
        start.record()
        fn(*args)
        end.record()
        torch.cuda.synchronize()
        samples.append(start.elapsed_time(end) * 1000)
    samples.sort()
    return {
        "median": statistics.median(samples),
        "p20": samples[int(len(samples) * 0.2)],
        "p80": samples[int(len(samples) * 0.8)],
    }

x = torch.randn(2048, 4096, dtype=torch.float16, device='cuda')
w = torch.ones(4096, dtype=torch.float16, device='cuda')
fn = lambda: torch.nn.functional.rms_norm(x, (4096,), w)

r1 = measure(fn, warmup=25, repeat=200)
r2 = measure(fn, warmup=25, repeat=200)

noise_pct = abs(r1["median"] - r2["median"]) / r1["median"] * 100
print(f"run1 median={r1['median']:.2f}µs  p20={r1['p20']:.2f}  p80={r1['p80']:.2f}")
print(f"run2 median={r2['median']:.2f}µs  p20={r2['p20']:.2f}  p80={r2['p80']:.2f}")
print(f"noise: {noise_pct:.2f}%")
```

- [ ] 上記を 5 回繰り返して median の最大ばらつき幅を記録

### 2. GPU クロック状態の影響確認

```bash
# クロックを固定する（可能な場合）
sudo nvidia-smi --lock-gpu-clocks=<max_clock>
# 計測後に解除
sudo nvidia-smi --reset-gpu-clocks
```

- [ ] クロック固定あり・なしでノイズレベルがどれだけ変わるか

### 3. 「3%」が検出可能な最小差分かを確認

```python
# PyTorch Eager vs 5% 速い仮想カーネルを 3% 閾値で検出できるか
```

---

## 記録すべき数値

| 項目 | 値 |
|---|---|
| GPU モデル | |
| median ばらつき（5 回の max-min / median） | |
| p80-p20 の幅（同一カーネル） | |
| クロック固定なしのノイズ率 | |
| クロック固定ありのノイズ率 | |
| 推奨 `min_speedup` 値 | |

---

## 判断が必要な点

| 問い | 選択肢 |
|---|---|
| `min_speedup=1.03` は妥当か | 実測ノイズ率の 2 倍程度が目安 |
| warmup=25, repeat=200 は十分か | ノイズが大きければ repeat を増やす |
| クロック固定を推奨設定にするか | ユーザー依存（sudo 必要）なので任意オプションにする |

---

## 完了条件

- [ ] ノイズレベルの実測値を記録
- [ ] `min_speedup` の推奨値を決定
- [ ] `ADR-004` にノイズ実測値を追記

**Comment by flipslidersand** [comment]
## 予備データ（#3 スモークテスト時に取得）

GTX 1080 / 同一 Triton カーネルを warmup=25, repeat=200 で 2 回測定:

```
run1: median=159.7us  p20=158.2  p80=168.9
run2: median≈同等
noise (median run1 vs run2): 0.70%
```

### 示唆
- 測定ノイズ ≈ **0.7%**（GTX 1080、クロック固定なし、デスクトップ環境）
- ADR-004 の `min_speedup=1.03`（3%）は **ノイズの約 4 倍** → 妥当な保守的閾値
- p80 (168.9) と p20 (158.2) の幅は median の ~6.7%。`p80 < p20/1.03` 判定は機能しそう

### Phase 4（ノイズ調査）本実施時のTODO
- [ ] 5 回繰り返して median の max-min 幅を確認
- [ ] X サーバー稼働中（GUI 使用中）のノイズ増加を確認（現在 GUI プロセスが GPU 443MB 使用中）
- [ ] クロック固定ありとの比較

**Comment by flipslidersand** [comment]
## 確定: min_speedup=1.03（3%）を採用

`benchmark/statistics.py::is_improvement` で `p80(candidate) < p20(baseline) / 1.03` を採用。
Orchestrator のデフォルト `min_speedup=1.03`。

実走で機能を確認: ノイズ内（~1%差）の複数候補は BEST 扱いされず、統計的に有意な改善のみ採用。
12 候補の探索で 165-168us の候補が並んだが、最初の 1 件のみ BEST に固定され、残りは
「改善なし」と正しく判定された（保守的判定が機能）。

#3 で測定したノイズ ~0.7% に対し 3% は約 4 倍のマージン。→ **本 Issue はクローズ可。**
